efi/preinstall/check_pcr7.go

```diff
@@ -0,0 +1,537 @@
+				}
+
+				// Make sure that one of the CA's used for verification so far
+				// is a trust anchor for one of the signatures on the image.
+				var foundSig *efi.WinCertificateAuthenticode
+				for _, cert := range result.UsedAuthorities {
+					for _, sig := range sigs {
+						if sig.CertLikelyTrustAnchor(cert) {
+							foundSig = sig
+							break
+						}
+					}
+					if foundSig != nil {
+						break
+					}
+				}
+				if foundSig == nil {
+					return nil, errors.New("OS initial boot loader was not verified by any X.509 certificate measured by any EV_EFI_VARIABLE_AUTHORITY event")
+				}
+
+				signer := foundSig.GetSigner()
+				switch signer.PublicKeyAlgorithm {
+				case x509.RSA:
+					pubKey, ok := signer.PublicKey.(*rsa.PublicKey)
+					if !ok {
+						return nil, errors.New("signer certificate for OS initial boot loader contains unsupported public key type")
+					}
+					if pubKey.N.BitLen() <= 1024 {
+						result.Flags |= secureBootIncludesWeakAlg
+					}
+				default:
+					return nil, errors.New("signer certificate for OS initial boot loader contains unsupported public key algorithm")
+				}
+
+				// This is the launch of the IBL. At this point, events are under control of the
+				// OS, so we stop checking, even though we may miss some events created by
+				// firmware via the LoadImage API.
+				break NextEvent
+			}
+
+			if ev.PCRIndex != internal_efi.SecureBootPolicyPCR {
+				// Not PCR7
+				continue NextEvent
+			}
+
+			switch ev.EventType {
+			case tcglog.EventTypeEFIVariableAuthority:
+				// Decode the verification event
+				data, ok := ev.Data.(*tcglog.EFIVariableData)
+				if !ok {
+					// if decoding failed, the resulting data is guaranteed to implement error.
+					return nil, fmt.Errorf("EV_EFI_VARIABLE_AUTHORITY event has wrong data format: %w", ev.Data.(error))
+				}
+
+				// As we're only checking events up to the launch of the IBL, we don't expect
+				// to see anything other than verification events from db here.
+				if data.VariableName != efi.ImageSecurityDatabaseGuid || data.UnicodeName != "db" {
+					return nil, fmt.Errorf("EV_EFI_VARIABLE_AUTHORITY event is not from db (got %s-%v)", data.UnicodeName, data.VariableName)
+				}
+
+				if phase == tcglogPhaseOSPresent {
+					// Compute the expected digest from the event data in the log and make
+					// sure it's consistent with the measured digest. We only do this for
+					// OS-present events because these are the ones we compute. Pre-OS
+					// events are just copied from the log.
+					seenOSPresentVerification = true
+					expectedDigest := tcglog.ComputeEFIVariableDataDigest(pcrAlg.GetHash(), data.UnicodeName, data.VariableName, data.VariableData)
+					if !bytes.Equal(ev.Digests[pcrAlg], expectedDigest) {
+						return nil, fmt.Errorf("event data inconsistent with %v event digest for EV_EFI_VARIABLE_AUTHORITY event (log digest:%#x, expected digest:%#x)", pcrAlg, ev.Digests[pcrAlg], expectedDigest)
+					}
+				}
+
```

Comment:
the check below is relevant both for tcglogPhasePreOSAfterMeasureSecureBootConfig and tcglogPhaseOSPresent: ?

Reply:
It is.

Reply:
I've split up the branches for tcglogPhasePreOSAfterMeasureSecureBootConfig and tcglogPhaseOSPresent entirely now.